test/Sema/availability.swift

```diff
@@ -27,35 +27,53 @@ func foo(x : NSUInteger) { // expected-error {{'NSUInteger' is unavailable: use
   let z3 : MyModule.Outer.NSUInteger = 42 // expected-error {{'NSUInteger' is unavailable: use 'UInt' instead}}  
 }
 
-// Test preventing overrides of unavailable methods.
+// Test preventing overrides (but allowing shadowing) of unavailable methods.
 class ClassWithUnavailable {
   @available(*, unavailable)
   func doNotOverride() {} // expected-note {{'doNotOverride()' has been explicitly marked unavailable here}}
 
   // FIXME: extraneous diagnostic here
   @available(*, unavailable)
-  init(int _: Int) {} // expected-note 2 {{'init(int:)' has been explicitly marked unavailable here}}
+  init(int _: Int) {} // expected-note 3 {{'init(int:)' has been explicitly marked unavailable here}}
 
   convenience init(otherInt: Int) {
     self.init(int: otherInt) // expected-error {{'init(int:)' is unavailable}}
   }
 
   @available(*, unavailable)
-  subscript (i: Int) -> Int { // expected-note{{'subscript(_:)' has been explicitly marked unavailable here}}
+  subscript (i: Int) -> Int { // expected-note 2 {{'subscript(_:)' has been explicitly marked unavailable here}}
     return i
   }
 }
 
 class ClassWithOverride : ClassWithUnavailable {
-  override func doNotOverride() {} // expected-error {{cannot override 'doNotOverride' which has been marked unavailable}}
+  override func doNotOverride() {}
+  // expected-error@-1 {{cannot override 'doNotOverride' which has been marked unavailable}}
+  // expected-note@-2 {{remove 'override' modifier to declare a new 'doNotOverride'}} {{3-12=}}
+
+  override init(int: Int) {}
+  // expected-error@-1 {{cannot override 'init' which has been marked unavailable}}
+  // expected-note@-2 {{remove 'override' modifier to declare a new 'init'}} {{3-12=}}
+
+  override subscript (i: Int) -> Int { return i }
+  // expected-error@-1 {{cannot override 'subscript' which has been marked unavailable}}
+  // expected-note@-2 {{remove 'override' modifier to declare a new 'subscript'}} {{3-12=}}
+}
+
+class ClassWithShadowing : ClassWithUnavailable {
+  func doNotOverride() {} // no-error
+  init(int: Int) {} // no-error
+  subscript (i: Int) -> Int { return i } // no-error
 }
```

Comment:
Could you add a test for `required init` (and maybe `convenience init`) as well? They doesn't need an `override` keyword.

Reply:
I'm not sure how much coverage of different combinations you're looking for, but I'll force-push a version with a couple more test cases.